### PR TITLE
Make it possible to evaluate stdlib

### PR DESCRIPTION
**Pull request by sbarzowski**
New features in parser and desugarer:
* Support for slices
* Support for methods
* Support for function sugar in locals
* Prevent crashes with obj/arr comprehensions

After this change evaluating stdlib doesn't crash.

### CODE REVIEW

**Comment by sbarzowski** [comment]
@sparkprime @lwander

File: parser.go [hunk 1]
```diff
@@ -944,18 +948,59 @@ func (p *parser) parse(prec precedence) (astNode, error) {
...
+						end = p.pop()
+						readyForNextIndex = true
+					} else if readyForNextIndex {
+						indexes[which], err = p.parse(maxPrecedence)
+						if err != nil {
+							return nil, err
+						}
+						readyForNextIndex = false
+					} else {
+						_, err := p.popExpect(tokenBracketR)
```
Comment (lwander) [inline parser.go]: I'm missing something here, shouldn't the fact that the first `if p.peek().kind == tokenBracketR` failed mean that this `p.popExpect(tokenBracketR)` will always fail?

File: parser_test.go [hunk 2]
```diff
@@ -200,6 +202,8 @@ var errorTests = []testError{
 
 	{`a[(b c)]`, `test:1:6-7 Expected token ")" but got (IDENTIFIER, "c")`},
 	{`a[b c]`, `test:1:5-6 Expected token "]" but got (IDENTIFIER, "c")`},
+	{`a[]`, `test:1:3-4 Index requires an expression`},
+	{`a[42:42:42:42]`, `test:1:11-12 Invalid slice: too many colons`},
```
Comment (lwander) [inline parser_test.go]: Some non-error tests would be helpful too

File: desugarer.go [hunk 3]
```diff
@@ -128,16 +128,21 @@ func desugarFields(location LocationRange, fields *astObjectFields, objLevel int
 
 	// Remove methods
 	// TODO(dcunnin): this
-	for _, field := range *fields {
+	for i := range *fields {
+		field := &((*fields)[i])
```
Comment (lwander) [inline desugarer.go]: Why take the address?

File: desugarer.go [hunk 4]
```diff
@@ -301,8 +335,33 @@ func desugar(astPtr *astNode, objLevel int) (err error) {
 	case *astIndex:
 		return unimplErr
 
+	case *astSlice:
+		stdslice := &astVar{id: "std.slice"}
+		var apply astNode
```
Comment (sbarzowski) [inline desugarer.go]: I need *astNode (a pointer to astNode) and this is a different type from **astVar. And I didn't know I can do `var x type = y` :-). 

done

File: parser.go (hunk 1 above)
Comment (sbarzowski) [inline parser.go]: Yeah. I'm just using this function to get the right error message. I have changed this to be more explicit.

File: parser_test.go (hunk 2 above)
Comment (sbarzowski) [inline parser_test.go]: You are right, I forgot. And I have found (and fixed) a bug - missing :: handling. Thank you!

File: desugarer.go (hunk 3 above)
Comment (sbarzowski) [inline desugarer.go]: B/c it's modified in place. I don't want to make a copy, b/c then the changes won't be made in the "real" ast.

Comment (lwander) [inline desugarer.go]: Ah I missed that, sounds good

File: parser.go [hunk 5]
```diff
@@ -944,18 +960,58 @@ func (p *parser) parse(prec precedence) (astNode, error) {
 			op := p.pop()
 			switch op.kind {
 			case tokenBracketL:
-				index, err := p.parse(maxPrecedence)
-				if err != nil {
-					return nil, err
+				// handle slice
+				var indexes [3]astNode
+				which := 0
```
Comment (sparkprime) [inline parser.go]: Maybe call it colons_consumed

**Comment by sparkprime** [comment]
BTW I noticed coverage is steadily decreasing.  Can you keep that in mind as we will have to root out untested code at some point anyway.

**Pull request merged**